File: src/AutomatonTheory/JavaLogic/DeterministicFiniteAutomaton.java
package AutomatonTheory.JavaLogic;

import java.util.List;

public class DeterministicFiniteAutomaton extends Automaton {

    public DeterministicFiniteAutomaton(String automatonName){
        AutomatonName = automatonName;
        Type = Automatons.DFA;
    }

    public DeterministicFiniteAutomaton(String automatonName, List<String> alphabet){
        this(automatonName);
        setAlphabet(alphabet);
    }

    @Override
    public boolean evaluateString(String evaluationString) {
        char[] characters = evaluationString.toCharArray();
        State currentState = getInitialState();

        if(characters.length > 0) {
            for (char character : characters) {
                if(Alphabet.contains(""+character)){
                    boolean hasTransition = false;
                    for(Transition item : currentState.Transitions){

                        if(item.Symbol.equals(""+character)){
                            currentState = item.DestinyState;
                            hasTransition = true;
                        }
                    }
                    if(!hasTransition){
                        return false;
                    }
                }
                else{
                    return false;
                }
            }
            if(currentState.AcceptanceState){
                return true;
            }
            else{
                return false;
            }
        }
        return false;
    }

    @Override
    public boolean addTransition(String originStateName, String destinyStateName, String symbol) {
        State originState = getState(originStateName);
        State destinyState = getState(destinyStateName);

        //validar que el origen y symbol no se pueden repetir
        boolean canAddTransition = true;
        for(State state : States){
            for(Transition transition : state.Transitions){
                if(state.Name.equals(originStateName) &&
                        transition.Symbol.equals(symbol)){
                    canAddTransition = false;
                }
            }
        }
        if(canAddTransition){
            return originState.addTransition(new Transition(destinyState, symbol));
        }
        return false;
    }
}

File: src/AutomatonTheory/Kotlin/AutomatonExtensions/DeterministicFiniteAutomaton.kt
package AutomatonTheory.Kotlin.AutomatonExtensions

import AutomatonTheory.Kotlin.AutomatonLogic.Automaton
import AutomatonTheory.Kotlin.AutomatonLogic.Automatons
import AutomatonTheory.Kotlin.AutomatonLogic.Transition

open class DeterministicFiniteAutomaton(automatonName: String) : Automaton() {

    init {
        AutomatonName = automatonName
        Type = Automatons.DFA
    }

    constructor(automatonName: String, alphabet: MutableList<String>) : this(automatonName) {
        setAlphabet(alphabet)
    }

    override fun evaluateString(evaluationString: String): Boolean {
        val characters = evaluationString.toCharArray()
        var currentState = getInitialState()

        if (characters.size > 0) {
            for (character in characters) {
                if (Alphabet.contains("" + character)) {
                    var hasTransition = false
                    for (item in currentState.Transitions) {

                        if (item.Symbol == "" + character) {
                            currentState = item.DestinyState
                            hasTransition = true
                        }
                    }
                    if (!hasTransition) {
                        return false
                    }
                } else {
                    return false
                }
            }
            if (currentState.AcceptanceState) {
                return true
            } else {
                return false
            }
        }
        return false
    }

    override fun addTransition(originStateName: String, destinyStateName: String, symbol: String): Boolean {
        val originState = getState(originStateName)
        val destinyState = getState(destinyStateName)

        //validar que el origen y symbol no se pueden repetir
        var canAddTransition = true
        for (state in States) {
            for (transition in state.Transitions) {
                if (state.Name == originStateName && transition.Symbol == symbol) {
                    canAddTransition = false
                    return false;
                }
            }
        }
        if (canAddTransition) {
            return originState.addTransition(Transition(destinyState, symbol))
        }
        return false
    }
}
